[2단계 - 출석 구현] 헤일러(민서현) 미션 제출합니다.

threepebbles opened the pull request:
안녕하세요 샐리
이번 단계에서는 혼자 고민도 많이 해보고 주변 크루들과도 많은 의논을 하며 진행해봤어요. 
리뷰 잘 부탁드립니다! 😄 

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [헤일러의 프롤로그](https://prolog.techcourse.co.kr/studylogs/3903)에 셀프 체크를 작성했나요?


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
* 메서드 분리, 클래스 분리를 만족스러울 만큼 해내지는 못한 것 같습니다.
* 원시값과 문자열을 적절하게 잘 포장했는지 모르겠습니다.

## ❗ 집중해서 리뷰 받고 싶은 내용 및 질문
### 0️⃣ switch/case
switch/case를 사용하지 말라는 요구사항이 있었으나, 메인 화면에서 기능 선택을 처리하는 로직은 if/else if보다 switch/case을 사용하는 것이 더 가독성이 좋다는 생각이 들어 switch/case를 사용했습니다.
같은 로직을 switch/case가 아닌 다른 방식으로 더 가독성 좋게 작성하는 방법이 있을까요?
```java
private boolean executeCommand(String command) {
    boolean isRunning = true;
    switch (MainMenuCommand.from(command)) {
        case SAVE_ATTENDANCE_RECORD -> saveAttendanceRecord();
        case MODIFY_ATTENDANCE_RECORD -> modifyAttendanceRecord();
        case PRINT_MONTH_ATTENDANCE_STATISTICS -> printMonthAttendanceStatistics();
        case PRINT_RISK_CREWS -> printRiskCrews();
        case QUIT, QUIT_SMALL_CASE -> isRunning = false;
        default -> System.out.println("존재하지 않는 커맨드입니다.");
    }
    return isRunning;
}
```

### 1️⃣ TDD 사이클 커밋에 대한 피드백
TDD를 잘 실천하고 있는지 기록을 남기기 위해 [red1: 닉네임으로 존재하는 크루인지 확인한다](https://github.com/woowacourse/java-attendance/commit/e5fd59b36f27a10585ec6a347f742431c62a0e8e)부터 [green2: 날짜와 시간으로 알맞는 출석 상태를 반환한다](https://github.com/woowacourse/java-attendance/commit/0b3ae056267cba74cd9c6866fee0ef6ea002daa0)까지는 red사이클과 green 사이클을 구분해서 커밋을 작성했습니다.
red 사이클에서 테스트 코드를 먼저 작성하며 선언되지 않은 변수들은 Problem-Solving 단축키(windows기준 alt+enter)로 내용이 빈 스켈레톤 코드를 생성해놓고 green 사이클을 진행했습니다. 
테스트 단위의 크기가 적절한 지 보완할 점은 없을 지 피드백 부탁드립니다!

### 2️⃣ 설계와 TDD
이번 미션의 목표가 TDD, 리팩터링인만큼 TDD 사이클을 최대한 지키도록 노력하면서 코드를 작성했습니다.
스스로 TDD를 진행해본 경험이 처음이다 보니 많은 우여곡절이 있었는데요 🥹 
가장 많이 부딪혔던 문제는 red 사이클에서 작성했던 테스트 함수를 만족시키기 위해 green 사이클에서 작성해야 하는 코드의 양이 너무 많은 경우였습니다.
이런 상황을 최대한 예방하고자 설계는 Top-down 방식으로, 구현은 Bottom-up 방식으로 하려고 노력했습니다.
제가 했던 Top-down 방식 설계는 아래와 같이 진행했습니다.
1. 기능 요구 사항 중 UI 로직을 제외한 **도메인 로직을** 분리하여 **정리**합니다.
2. 도메인 로직을 하나의 큰 흐름 단위로 묶어서 분류합니다.
ex) 출석 기록 저장 기능, 출석 기록 수정 기능, 크루의 월단위 출석 기록 조회 기능, 제적 위험자 목록 조회 기능 각각이 하나의 큰 흐름
3. 각각의 큰 흐름을 구현하기 위해 필요한 기능들을 세부 기능으로 분리합니다. 제가 설계할 수 있는 한도 내에서 최대한 가장 작은 단위의 기능으로 나누어질 때까지 기능을 작게 쪼갭니다.
```
ex)
- 출석 기록 저장 기능
  - 존재하는 크루인지 검증한다.
  - 출석 기록을 저장한다.
    - 출석 기록 = 크루의 닉네임, 출석 날짜(연월일), 출석 시간(시분), 출석 상태(출석, 지각, 결석) 
- 출석 기록 수정 기능
  - 존재하는 크루인지 검증한다.
  - 출석 기록을 수정한다.
- 크루의 월단위 출석 기록 통계를 조회한다.
  - 닉네임으로 해당 크루의 월단위 출석 기록을 조회한다.
  - 해당 출석 기록의 출석 상태별 횟수를 계산한다.
  - 출석 상태별 횟수를 기반으로 제적 위험 등급을 계산한다.
- 제적 위험자 목록 조회 기능
  - 모든 크루 목록을 조회한다.
  - 각 크루별 출석 상태별 횟수를 계산한다.
  - 출석 상태별 횟수를 기반으로 제적 위험 등급을 계산한다.
```
4. 위의 설계를 기반으로 가장 작은 기능 단위부터 TDD의 red-green 사이클을 시작합니다.

하지만 아직 설계 경험의 부족으로 인해 나눈 기능의 단위가 생각보다 커서 하나의 red 사이클을 만족시키기 위해 너무 많은 양의 코드가 작성되는 경우가 많았습니다.
그렇다고 많은 양의 코드를 한번에 green 사이클에 포함시키게 되면 새로 추가된 많은 양의 코드들에 대한 테스트가 부족할 것이기 때문에 새로운 진행하던 green 사이클을 멈추고, 새로운 red 사이클을 만들어서 진행하는 방식으로 했습니다.

![tdd drawio](https://github.com/user-attachments/assets/0d47effa-ce85-4c32-a757-d50494f29e3f)
위의 그림으로 예시를 들면 red2 테스트를 만족시키는 프로덕션 코드를 작성하던 도중, red2가 아닌 다른 테스트가 필요한 프로덕션 코드들이 생긴 경우, green2를 멈추고 red3-green3, red4-green4, red5-green5를 먼저 진행합니다. 그리고 green5까지 끝난 후 green2로 다시 돌아와서 red2를 통과시키는 방식으로 진행했습니다.

샐리는 green사이클에서 너무 많은 양의 프로덕션 코드를 작성하게 됐을 때 어떻게 대응하시나요? 샐리가 설계와 TDD를 하는 방식도 궁금합니다!

### 3️⃣ Service 계층과 View 계층의 완전한(?) 분리
View <-> Controller <-> Service <-> Domain (+ Repository)
우선 계층은 MVC패턴을 기반으로 위와 같이 4개의 계층으로 분리했습니다.
그리고 아래의 2가지 기준은 꼭 지키도록 노력했습니다.
1. Service는 Domain에만 의존한다.
2. View는 Domain 정보를 알 수 없다.

2번 규칙을 지키기 위해 dto를 적극적으로 사용했고, Service 계층에서 Controller 계층으로 반환하는 모든 값은 primitive 값 혹은 POJO만 사용하게 했습니다. 그러다보니 출석 시간(월일, 시분), 출석 상태(출석, 지각, 결석), 제적 위험 등급(경고, 면담, 제적)과 같은 정보도 다 String 타입으로 넘겼습니다.
제가 너무 엄격하게 격리한 것은 아닌지, enum 클래스(`AttendanceStatus`, `RiskRank`)정도는 View가 알아도 상관없는지 여쭈어보고 싶습니다!

### 4️⃣ 논리적으로 도달할 수 없는 예외에 대한 처리
```java
public enum RiskRank {
    NOT_MANAGED("", 0),
    WARNING("경고", 2),
    INTERVIEW("면담", 3),
    EXPELLED("제적", 6),
    ;
    ...

    public static RiskRank from(int riskCount) {
        if (riskCount < 0) {
            throw new IllegalArgumentException(riskCount + ": 결석 횟수는 음수일 수 없습니다.");
        }

        return Arrays.stream(values())
                .filter(riskRank -> riskRank.riskCountLimit <= riskCount)
                .max(Comparator.comparing(RiskRank::getRiskCountLimit))
                .orElseThrow(() -> new IllegalStateException("논리적으로 도달할 수 없는 예외입니다."));
    }
}
```
from 메서드는 결석 횟수를 파라미터로 받아 그에 맞는 제적 위험 등급 enum을 반환하는 메서드입니다.
riskCount가 음수인 경우를 제외하고는 orElseThrow까지 논리적으로 도달할 수 없는 상황입니다.
이런 상황에서 `IllegalStateException`을 호출하는 것이 자연스러운지, 아니면 더 의도적으로 메세지를 잘 전달할 방법이 있는지 궁금합니다.

### 5️⃣ 기획 요구 사항이 일관되지 않은 형식(?)으로 주어졌을 때, 새로운 일관된 기준을 정의하고 적용
```java
public enum RiskRank {
    NOT_MANAGED("", 0),
    WARNING("경고", 2),
    INTERVIEW("면담", 3),
    EXPELLED("제적", 6),
    ;

    private final String name;
    private final int riskCountLimit;

    ...
    public static AttendanceStatus of(LocalDate date, LocalTime time) {
        if (!LectureTime.isLectureDate(date)) {
            throw new IllegalArgumentException(date + ": 교육이 없는 날입니다.");
        }

        int elapsedMinutes = LectureTime.calculateElapsedMinutes(date, time);
        return Arrays.stream(values())
                .filter(status -> status.elapsedMinutesLimit < elapsedMinutes)
                .max(Comparator.comparing(AttendanceStatus::getElapsedMinutesLimit))
                .orElseThrow(() -> new IllegalStateException("논리적으로 발생할 수 없는 예외입니다."));
    }
```
기획 요구 사항에서는 결석 2회 **이상**이 경고 대상자, 결석 3회 **이상**이 면담 대상자, 결석 5회 **초과**가 제적 대상자라고 주어졌습니다. **이상**과 **초과**는 서로 다른 기준이기 때문에 5회 초과를 6회 이상으로 새로 해석하여 **이상**을 기준으로 상수값을 정했습니다. 결국 기획에서 요구한 값과 다른 상수 값을 코드 내에서 사용하게 되는데 이렇게 해도 괜찮은지 여쭤보고 싶습니다!

### 6️⃣ `ATTENDANCE`와 `ABSENT`의 elapsedMinutesLimit(늦은 시간(단위:분)) 값
```java
public enum AttendanceStatus {
    ATTENDANCE("출석", "정상 등교", -1),
    LATE("지각", "교육 시작 시간 5분 초과", 5),
    ABSENT_LATE("결석", "교육 시작 시간 30분 초과", 30),
    ABSENT("결석", "출석 기록 없음", Integer.MAX_VALUE),
    ;

    private String name;
    private String description;
    private int elapsedMinutesLimit;

    ...
    public static AttendanceStatus of(LocalDate date, LocalTime time) {
        if (!LectureTime.isLectureDate(date)) {
            throw new IllegalArgumentException(date + ": 교육이 없는 날입니다.");
        }

        int elapsedMinutes = LectureTime.calculateElapsedMinutes(date, time);
        return Arrays.stream(values())
                .filter(status -> status.elapsedMinutesLimit < elapsedMinutes)
                .max(Comparator.comparing(AttendanceStatus::getElapsedMinutesLimit))
                .orElseThrow(() -> new IllegalStateException("논리적으로 발생할 수 없는 예외입니다."));
    }
```
출석 상태라는 관점에서 출석, 지각, 30분 초과 결석, 출석 기록 없는 결석은 하나의 관심사로 묶일 수 있습니다.
그런데 `ATTENDANCE`와 `ABSENT`는 늦은 시간(elapsedMinutesLimit)을 정의할 수 없습니다. 
날짜(연월일)와 시간(시분) 정보를 받아 알맞는 출석 상태를 반환하는 `of` 메서드를 편하게 사용하기 위해 임의의 값인 -1과 Integer.MAX_VALUE를 사용했는데, 이는 다소 해석이 모호할 수 있는 값이란 생각이 듭니다.
이런 경우 출석 상태라는 도메인을 묶어낼 수 있는 더 좋은 방법이 있을지 여쭤보고 싶습니다.

threepebbles committed:
remove: step1 코드 삭제

threepebbles committed:
docs: 기능 설계서 재작성

threepebbles committed:
red1: 닉네임으로 존재하는 크루인지 확인한다

threepebbles committed:
green1: 닉네임으로 존재하는 크루인지 확인한다

threepebbles committed:
red2: 날짜와 시간으로 알맞는 출석 상태를 반환한다

threepebbles committed:
red3: 해당 날짜가 주말인지 확인한다

threepebbles committed:
green3: 해당 날짜가 주말인지 확인한다

threepebbles committed:
red4: 해당 날짜가 공휴일인지 확인한다

threepebbles committed:
green4: 해당 날짜가 공휴일인지 확인한다

threepebbles committed:
refactor(AttendanceStatus): 패키치 이동

threepebbles committed:
red5: 해당 날짜가 교육이 있는 날인지 확인한다

threepebbles committed:
green5: 해당 날짜가 교육이 있는 날인지 확인한다

threepebbles committed:
refactor(AttendanceStatusTest): 패키지 이동

threepebbles committed:
red6: 해당 날짜에 맞는 교육 시간을 계산한다

threepebbles committed:
green6: 해당 날짜에 맞는 교육 시간을 계산한다

threepebbles committed:
red7,green7: 날짜와 시간으로 교육 시작 시각으로부터 몇분 늦었는지 계산한다

threepebbles committed:
green2: 날짜와 시간으로 알맞는 출석 상태를 반환한다

threepebbles committed:
이후로 red, green 사이클을 묶어서 feat으로 커밋

threepebbles committed:
feat: 날짜(월일)에 특정 크루의 출석 기록이 있는지 확인하는 기능 추가

threepebbles committed:
feat: attendances.csv로부터 출석 기록을 읽어오는 기능 추가

threepebbles committed:
feat(InputView): 사용자로부터 닉네임 입력받는 기능 추가

threepebbles committed:
feat(InputView): 사용자로부터 등교 시간 입력 받는 기능 추가

threepebbles committed:
feat(OutputView): 저장한 출석 기록 출력하는 기능 추가

threepebbles committed:
feat: 메인 화면 안내 기능 추가

threepebbles committed:
feat(AttendanceService): 닉네임과 등교시간을 받아 출석 기록을 저장하는 기능 추가

threepebbles committed:
test: 출석 기록 수정 기능 테스트 추가

threepebbles committed:
feat(AttendanceService): 출석 기록 수정 기능 추가

threepebbles committed:
test: 수정하려는 출석 기록과 동일한 출석 기록이 존재하는 경우 예외 테스트 추가

threepebbles committed:
docs: 구현된 기능 목록 체크

threepebbles committed:
refactor: 주말 및 공휴일 검증 기능 선언 클래스 변경

threepebbles committed:
refactor: 메서드명 변경

threepebbles committed:
docs: 출석을 기록할 수 없는 기준 따로 분류

threepebbles committed:
docs: 구현된 기능 목록 체크

threepebbles committed:
refactor: 패키지 이동

threepebbles committed:
test: 크루별 출석 기록 확인 기능 테스트 추가

threepebbles committed:
fix(LectureTime): 교육이 있는 날짜인지 검증하는 메서드에 주말 및 공휴일 검증 추가

threepebbles committed:
feat(AttendanceService): 특정 크루의 월 단위 출석 기록을 불러오는 기능 추가

threepebbles committed:
refactor(AttendanceService): getMonthAttendanceRecords 메서드 분리

threepebbles committed:
docs: 크루별 출석 기록 확인 기능 입력 예제 추가

threepebbles committed:
feat: 출석 기록 수정 후 변경 전과 변경 후 출석 기록 출력 기능 추가

threepebbles committed:
refactor: 메서드 파라미터 변경

threepebbles committed:
remove: 사용하지 않는 메서드 제거

threepebbles committed:
feat(AttendanceService): 제적 위험자 목록 추출 기능 추가

threepebbles committed:
feat: 출석 기록 통계 출력 기능 추가

threepebbles committed:
feat: 제적 위험자 목록 출력 기능 추가

threepebbles committed:
refactor: service.dto 패키지 하위 response 객체는 모두 원시값 혹은 POJO 객체만 필드로 갖도록 변경

threepebbles committed:
refactor: 의도가 드러나도록 변수명 수정

threepebbles committed:
test: 테스트명 수정

threepebbles committed:
feat(DateTimeUtil#isInRange): 시작 시간이 종료 시간보다 뒤인 경우 검증 추가

threepebbles committed:
feat: 예외 발생 시, 프로그램이 종료되지 않고 메인 화면으로 돌아감

threepebbles committed:
feat: controller.dto 패키지 하위 dto들에 대해 생성자 검증 추가

threepebbles committed:
refactor: 변수명 일관되게 수정

threepebbles committed:
test: 매 테스트 실행 시 Crew 저장소 초기화

threepebbles committed:
refactor: 테스트 메서드명 일관되게 변경

threepebbles committed:
test: 크루의 출석 기록이 존재하지 않는 경우 테스트 추가

threepebbles committed:
refactor: 메서드 분리

threepebbles committed:
fix: 존재하지 않는 커맨드 입력 시 종료되지 않고 커맨드 재입력 받도록 수정

threepebbles committed:
refactor: 변수명 변경(option -> command)

threepebbles committed:
refactor: 출석 상태별 횟수를 저장하는 자료구조 변경

threepebbles committed:
chore: attendances.csv 데이터 수정

threepebbles committed:
refactor(RiskCrew): 제적 위험자 목록 저장하는 자료구조 변경(List -> TreeSet)

threepebbles committed:
refactor: 접근 제어자 수정

threepebbles committed:
refactor: 공백 수정

threepebbles committed:
refactor: 메서드 분리

threepebbles committed:
refactor: 네이밍 변경

threepebbles committed:
fix(AttendanceRecordRepository): 싱글턴으로 사용하기 위해 private 생성자 추가

threepebbles committed:
remove: 주석 제거

threepebbles committed:
refactor(ExceptionHandler): 메서드 분리

threepebbles committed:
docs: 구현한 기능 목록 체크

threepebbles committed:
feat(AttendanceStatus): 중복된 검증 제거

threepebbles committed:
docs: 설명 구체화

threepebbles committed:
fix(SaveAttendanceRecordResponse): 필드로 도메인 객체 대신 원시 변수를 가지도록 타입 수정

threepebbles committed:
test: 결석 횟수에 맞는 제적 위험 등급을 반환하는지 검증하는 테스트 추가

threepebbles committed:
feat: 메인 화면에 q커맨드(Q커맨드와 동일) 추가

threepebbles committed:
refactor(AttendanceStatus): 상수값 수정

threepebbles committed:
fix: 출석 기록 저장소 초기화 시기 수정(@BeforeEach -> @AfterEach)

threepebbles committed:
feat(AttendanceService): 크루별 월단위 출석 기록 확인 기능에서 존재하는 크루인지 검증하도록 수정

threepebbles commented:
리뷰 요청 후에 2개 커밋 추가했습니다..! 죄송해요

threepebbles committed:
refactor: 상수 네이밍 컨벤션에 맞게 수정

threepebbles commented:
리뷰 요청 후에 3개 커밋 추가했습니다..! 죄송해요
크지 않은 변경사항입니다

threepebbles commented on src/main/java/service/dto/RiskCrew.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public int compareTo(RiskCrew o) {
+        int riskCount = riskCount();
+        int opponentRiskCount = o.riskCount();
+        if (riskCount == opponentRiskCount) {
+            return nickname.compareTo(o.nickname);
+        }
+        return o.riskCount() - riskCount();
```
메서드 두번 호출 안하려고 riskCount와 opponentRiskCount 변수를 선언한건데 마지막에 수정하는 걸 깜빡했네요 🥹 
피드백 주신 뒤에 수정하겠습니다.
```java
return opponentRiskCount - riskCount
```

dusdn1702 replied:
넵~ comparable 을 사용하신 이유도 궁금해요!

threepebbles replied:
RiskCrews 클래스의 TreeSet<RiskCrew>에서 정렬기준을 정의해주기 위해 Comparable을 사용했습니다!

dusdn1702 commented on src/main/java/controller/dto/ModifyAttendanceRequest.java:
```diff
@@ -0,0 +1,38 @@
+
+    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm");
+
+    public static ModifyAttendanceRequest of(String nickname, LocalDate today, int day, String time) {
```
이 함수는 객체 생성 외에 유효성 검증도 진행하는 것 같은데 이름을 바꾸면 어떨까요?

dusdn1702 commented on src/main/java/controller/dto/ModifyAttendanceRequest.java:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    private static void validateTime(String time) {
```
유효성 검증을 dto 에서 해주신 이유가 궁금해요~
dto 는 무슨 역할인가요? 특히나 입력에 대한 파싱은 어디서 처리하면 좋을까요?

threepebbles replied:
조금 더 고민해보겠습니다!

dusdn1702 commented on src/main/java/controller/dto/ModifyAttendanceRequest.java:
```diff
@@ -0,0 +1,38 @@
+
+public record ModifyAttendanceRequest(
+        String nickname,
+        LocalDate date,
```
무슨 date이고 무슨 time인지 상세한 변수명을 이용하면 좋겠어요.
변경하고 싶은 date인지 기존 date인지 모르겠네요!

dusdn1702 commented on src/main/java/controller/dto/SaveAttendanceRequest.java:
```diff
@@ -0,0 +1,28 @@
+        LocalTime time
+) {
+
+    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm");
```
동일한 TIME_FORMAT을 request마다 선언해줄 필요가 있을까요?
한군데서 관리한다면 어디가 좋을까요?

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,44 @@
-    private final String title;
-    private final String description;
-    private final int elapsedMinutesLimit;
+    private String name;
```
enum은 LATE.name()으로 LATE 값이 사용되기도 하는데요.
getName()과 혼란이 있겠네요.

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,44 @@
+    public static AttendanceStatus of(LocalDate date, LocalTime time) {
+        if (!LectureTime.isLectureDate(date)) {
+            throw new IllegalArgumentException(date + ": 교육이 없는 날입니다.");
+        }
```
함수로 분리하지 않은 이유가 있을까요?

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,44 @@
+    ATTENDANCE("출석", "정상 등교", -1),
+    LATE("지각", "교육 시작 시간 5분 초과", 5),
+    ABSENT_LATE("결석", "교육 시작 시간 30분 초과", 30),
+    ABSENT("결석", "출석 기록 없음", Integer.MAX_VALUE),
```
같은 결석을 분리해서 관리할 필요가 있을까요?

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,44 @@
+                .filter(status -> status.elapsedMinutesLimit < elapsedMinutes)
                 .max(Comparator.comparing(AttendanceStatus::getElapsedMinutesLimit))
-                .orElse(AttendanceStatus.OFF_DAY);
+                .orElseThrow(() -> new IllegalStateException("논리적으로 발생할 수 없는 예외입니다."));
```
이것도 어떤 행위를 하는지 함수로 분리해서 명시하면 좋을 것 같아요.

dusdn1702 commented on src/main/java/domain/CampusTime.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.time.LocalTime;
+
+public class CampusTime {
```
아무런 함수도 가지지 않았다는 것은 역할이 없는 것일텐데요.
역할이 없는데 domain 패키지에 class 로 존재해야할 이유가 있을까요?

dusdn1702 commented on src/main/java/domain/Holiday.java:
```diff
@@ -5,23 +5,24 @@
+                .anyMatch(holiday -> holiday.month == month && holiday.day == day);
+    }
+
+    public static boolean isHoliday(LocalDate date) {
```
호출처가 없네요!

dusdn1702 commented on src/main/java/domain/Holiday.java:
```diff
@@ -5,23 +5,24 @@
     }
 
-    public static Holiday from(LocalDate date) {
+    public static boolean isHoliday(int month, int day) {
```
holiday임을 판단하는게 도메인 로직이라고 생각하신 이유가 궁금해요.
휴일 Holiday가 도메인인걸까요?

dusdn1702 commented on src/main/java/domain/LectureTime.java:
```diff
@@ -23,18 +26,30 @@ public enum LectureTime {
-    public LocalTime getEndTime() {
-        return endTime;
+    public static int calculateElapsedMinutes(LocalDate date, LocalTime time) {
+        if (!isLectureDate(date)) {
+            throw new IllegalArgumentException(date + ": 교육이 없는 날입니다.");
+        }
```
하나의 행위만 하는게 좋을 것 같아요.
calculateElapsedMinutes 호출 전에 해당 로직을 이미 수행하고 있네요.

dusdn1702 commented on src/main/java/domain/RiskRank.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static RiskRank from(int riskCount) {
+        if (riskCount < 0) {
+            throw new IllegalArgumentException(riskCount + ": 결석 횟수는 음수일 수 없습니다.");
+        }
```
음수는 계산해서 넘길 때 체크하는게 좋지 않을까요?

dusdn1702 commented on src/main/java/repository/AttendanceRecordRepository.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.List;
+
+public class AttendanceRecordRepository {
+    private final static List<AttendanceRecord> ATTENDANCE_RECORDS = new ArrayList<>();
```
ATTENDANCE_RECORDS 가 상수인가요? 이렇게 선언하면 상수처럼 인지하는 개발자가 많을텐데 괜찮을까요?

threepebbles replied:
static final 타입은 항상 upper snake case를 사용했는데 이런 역할의 변수는 다른 네이밍 컨벤션을 사용하나요?

dusdn1702 replied:
https://github.com/woowacourse/java-attendance/pull/84#discussion_r1972775612 
요 코멘트를 적용하면 static final 을 사용하지 않아도 되지 않을까요?
static final 은 어떠한 경우에 사용하시나요? 우리가 대문자로 규칙을 지어 사용하는 이유가 무엇일까요?

threepebbles replied:
피드백 주셔서 곰곰히 생각해봤는데 지금 미션에서는 파일에 데이터를 저장해서 영구히 저장하는 방식을 사용하고 있지 않기 때문에 전역 변수(static final)로 모든 곳에서 접근 가능하게 열어 놓는 방식보다는 의존성 주입을 통해 관리하는 것이 더 계층 분리가 자연스럽게 이루어질 것 같아요! 말씀해주신대로 리팩터링해보고 있습니다.

혹시 그것과 별개로 `static final List<T>`와 같이 primitive 타입이 아닌데 static final 키워드를 붙인 경우 상수라고 취급할 수 없을까요?

dusdn1702 replied:
static final 키워드를 붙인 경우 상수라고 취급할 수 없을까요? 요걸 잘 이해하지 못했어요! 어떤 의미인가요?

dusdn1702 commented on src/main/java/repository/AttendanceRecordRepository.java:
```diff
@@ -0,0 +1,55 @@
+public class AttendanceRecordRepository {
+    private final static List<AttendanceRecord> ATTENDANCE_RECORDS = new ArrayList<>();
+
+    private AttendanceRecordRepository() {
```
저장소란 의미의 repository 를 사용할거라면, static이 아닌 프로그램 실행 시에 하나의 저장소 (List<AttendanceRecord>)를 생성하고 객체로 관리하는게 맞지 않을까요?

dusdn1702 commented on src/main/java/repository/AttendanceRecordRepository.java:
```diff
@@ -0,0 +1,55 @@
+    private AttendanceRecordRepository() {
+    }
+
+    public static void put(AttendanceRecord attendanceRecord) {
+        if (exists(attendanceRecord.nickname(), attendanceRecord.date())) {
+            ATTENDANCE_RECORDS.removeIf(record ->
+                    attendanceRecord.nickname().equals(record.nickname())
+                            && attendanceRecord.date().equals(record.date()));
+        }
```
이 객체가 가진 자료구조는 List인데 put을 이용하면 제거하고 추가될 것이라고 기대할 수 있을까요? 자세한 함수명을 가지면 좋을 것 같아요.

threepebbles replied:
더 좋은 네이밍 고민해보겠습니다!

dusdn1702 commented on src/main/java/repository/AttendanceRecordRepository.java:
```diff
@@ -0,0 +1,55 @@
+                .orElseThrow(() -> new IllegalArgumentException("출석 기록이 존재하지 않습니다."));
+    }
+
+    public static boolean exists(String nickname, LocalDate date) {
+        return ATTENDANCE_RECORDS.stream()
+                .anyMatch(record -> nickname.equals(record.nickname())
+                        && date.equals(record.date()));
+    }
+
+    public static boolean exists(String nickname, LocalDate date, LocalTime time) {
```
제가 이 출석 프로젝트를 함께 리팩터링하는 개발자로 투입이 된다면 어떤 함수를 어떨 때 써야하는지, 어떤 변수를 넣어야 할지 파악하기 어려워 보여요.
전체적으로 함수명 변수명 모두 자세히 작성하는 것이 좋을 것 같네요.

threepebbles replied:
exists의 검색 기준이 nickname하나면 existsByNickname 이런식으로 하고 싶었는데, 검색 기준이 많다보니 그렇게 하지 못했습니다.
파라미터 시그니처만으로 의도를 드러내려고 했는데 의도가 잘 전달이 되지 않은 것 같네요.
predicate을 파라미터로 넣어주게 되면 좀 더 의미가 명확해질 수 있을까요?

dusdn1702 commented on src/main/java/repository/CrewRepository.java:
```diff
@@ -2,36 +2,29 @@
+import java.util.Collections;
 import java.util.List;
 
 public class CrewRepository {
```
이 repository로 마찬가지입니다.
저장하고 있지 않은데 저장소인 repository가 맞을까요? 
일급컬렉션 Crews 도메인 역할을 수행하는 것 아닌지 고민해보면 좋을 것 같아요.

dusdn1702 commented on src/main/java/service/AttendanceRecordLoader.java:
```diff
@@ -0,0 +1,44 @@
+import repository.AttendanceRecordRepository;
+import repository.CrewRepository;
+
+public class AttendanceRecordLoader {
```
이 파일 파싱을 하는 객체가 service 패키지에 존재하는 이유가 있을까요?

threepebbles replied:
도메인 정보(출석 기록이자 크루 목록)이 담긴 파일을 읽어서 도메인 객체를 생성하고 저장하는 일을 하고 있기 때문에 Service 계층의 역할이라고 봤습니다.

dusdn1702 commented on src/main/java/service/AttendanceRecordLoader.java:
```diff
@@ -0,0 +1,44 @@
+        LocalDateTime d = LocalDateTime.parse(dateTime, DATETIME_FORMAT);
+        CrewRepository.addCrew(new Crew(nickname));
+        AttendanceRecordRepository.add(new AttendanceRecord(nickname, d.toLocalDate(), d.toLocalTime(),
+                AttendanceStatus.of(d.toLocalDate(), d.toLocalTime())));
```
파싱을 하는 view와 domain 객체 생성이 매우 긴밀하게 의존하고 있네요. 개선이 필요해요.

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,173 @@
+import service.dto.SaveAttendanceRecordResponse;
 import util.DateTimeUtil;
 
 public class AttendanceService {
```
이 서비스에서 수행하는 함수가 너무 많아요.
각 객체에 알맞는 함수(행위)를 분배하는 것이 좋을 것 같아요!

dusdn1702 replied:
많이 깔끔해졌네요 💯

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,173 @@
-        TimeAttendanceStatus before = TimeAttendanceStatus.of(crew, request.date());
-        crew.modifyAttendanceTime(request.date(), request.time());
-        TimeAttendanceStatus after = TimeAttendanceStatus.of(crew, request.date());
+        TimeStatus before = getTimeStatus(request.nickname(), request.date());
```
Status라는 객체명이라 response 내부 dto인줄 몰랐어요. 함수 해석이 어려워보여요~!

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,44 @@
-    ABSENT_LATE("결석", "출석 시간 30분 초과해서 결석", 30),
-    ABSENT("결석", "출석 기록을 하지 않아서 결석", Integer.MAX_VALUE),
-    OFF_DAY("쉬는 날", "출석 상태를 정의할 수 없음", Integer.MAX_VALUE),
+    ATTENDANCE("출석", "정상 등교", -1),
```
description 사용처가 없는데 enum 이름으로 표현해보는 건 어떨까요?

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,173 @@
-                            crew.getNickname(), attendanceStatusStatistics, Manage.of(attendanceStatusStatistics));
-                })
-                .toList();
+    private List<AttendanceRecordResponse> getMonthAttendanceRecordResponses(
```
get~~responses 라는 이름을 가진 함수 내부에서 계산이 이루어지는게 어색하네요.

threepebbles replied:
도메인에서 출석 기록을 의미하는 객체로 AttendanceRecord를 이미 사용하고 있어서 dto로 포장된 출석 기록에 대한 네이밍을 어떻게 할지 고민하다가 AttendancecRecordResponse로 지었습니다. 이게 아무래도 잘못된 네이밍 같은데.. 뾰족한 방법이 안 떠오릅니다. 
도메인 클래스 이름과 dto 클래스 이름이 겹치는 경우 domain.AttendanceRecord, service.dto.AttendanceRecord 이런식으로 동일하게 클래스명을 사용해도 괜찮을까요?

dusdn1702 replied:
좋은 방법 같진 않아요. AttendanceRecord 를 호출할 때 실수하기 너무 쉽겠죠??
AttendancecRecordResponse 이름은 좋은데, getMonthAttendanceRecordResponses라면 단순 매핑을 통해 response를 가져와야한다고 생각해요.
하나의 역할을 수행하는 함수가 되도록 출석을 확인하는 도메인 로직과, 매핑 로직을 분리해보면 좋을 것 같아요.

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,173 @@
+    }
+
+    private AttendanceStatusCount calculateAttendanceStatusCount(
+            List<AttendanceRecordResponse> monthAttendanceRecords) {
```
List<AttendanceRecordResponse>를 일급컬렉션으로 분리하고 계산하는 함수를 가져가면 좋을 듯 해요~

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,173 @@
+        int riskCount = attendanceStatusCount.lateCount() / 3
+                + attendanceStatusCount.absentCount();
+        return RiskRank.from(riskCount);
+    }
```
이것도 AttendanceStatusCount의 역할으로 보여요~

dusdn1702 commented on src/main/java/service/dto/ModifyAttendanceRecordResponse.java:
```diff
@@ -0,0 +1,32 @@
+        return new ModifyAttendanceRecordResponse(date.format(DATE_FORMAT), before, after);
+    }
+
+    public record TimeStatus(
```
내부 레코드를 사용하신 이유가 무엇일까요?
객체명도 어색하네요~

threepebbles replied:
아무래도 출력 값의 영향을 많이 받은 구조인데요.
{date} {before} -> {after} 형태의 출력을 필요로 해서 출석 시간과 출석 상태를 하나로 묶어서 표현했습니다.
```
{12월 03일 화요일} {10:07 (지각)} -> {09:58 (출석)} 수정 완료!
{date} {before} -> {after} 
```
(출석 시간, 출석 상태)묶음은 수정 기능에 대한 응답 값 내부에서만 사용되는 묶음이고, 외부로 클래스가 분리되어 있으면 응집도가 떨어진다고 생각하여 내부 클래스를 사용했습니다. 
(출석 시간, 출석 상태)라는 묶음을 하나의 단어로 정의할 수 없어서 네이밍은 일단 두 필드명의 연속으로 했습니다..! 클래스 구조가 많이 어색한가요?

dusdn1702 replied:
출석 출력을 위한 객체라면, 그에 맞는 이름을 가져야한다고 생각해요.
TimeStatus라는 이름을 보고 모두 헤일러의 의도를 파악할 수 있을까요?
저는 TimeStatus 의 이름을 보고 AttendanceStatus 처럼 시간에 대한 상태를 가진 enum일 것이라 생각했어요.

dusdn1702 commented on src/main/java/service/dto/RiskCrew.java:
```diff
@@ -0,0 +1,23 @@
+package service.dto;
+
+public record RiskCrew(
```
요 객체는 왜 dto에 있을까요?

threepebbles replied:
제적 위험자 목록 조회 기능의 응답 값으로 제적 위험 크루의 정보를 담고 있는 dto로 사용했습니다!

dusdn1702 commented on src/main/java/service/dto/RiskCrew.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public int riskCount() {
+        return lateCount + absentCount * 3;
```
3이라는 숫자도 상수화 하면 좋을 것 같아요

dusdn1702 commented on src/main/java/service/dto/RiskCrewsResponse.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.TreeSet;
+
+public record RiskCrewsResponse(
+        TreeSet<RiskCrew> riskCrews
```
TreeSet을 사용할 때의 장점이 무엇이었나요~?

threepebbles replied:
처음엔 List 자료구조를 사용했었는데요, List로 정렬을 하려고 하면 정렬을 할 위치(sort를 실행할 지점)와 계층을 정해야 하고, 정렬 기준을 어디에 정의하여 표현할 것인지에 대해서도 고민해야 합니다. (compareTo를 재정의할지, Comparable을 상속한 람다 메서드를 사용할지, thenComparing()을 이용해서 정렬 기준이 여러 개임을 명시적으로 드러낼지)

TreeSet 자료구조를 사용함으로써 프로덕션 코드도 간결해지고, 정렬할 위치에 대한 고민을 줄여주고, 어디서 정렬의 기준을 찾아야 할지 의도적으로 드러낼 수 있다는 것이 장점이란 생각이 들었습니다.

n개의 원소가 삽입되어야 한다고 했을 때, 
List 자료구조를 사용하는 경우, 모든 원소를 삽입하는데 O(n), 한번의 정렬을 하는데 O(nlog(n))의 시간복잡도를 가집니다. TreeSet 자료구조를 사용하는 경우, 모든 원소를 삽입하는데 O(nlog(n)), 출력 시 순회하면서 O(nlogn)의 시간복잡도를 가집니다.
List를 사용하나 TreeSet을 사용하나 둘 다 시간복잡도는 최종적으로 O(nlog(n))으로 성능에도 크게 차이가 없기 때문에 위의 장점을 갖고 있는 TreeSet을 사용했습니다.

dusdn1702 commented on src/main/java/util/DateTimeUtil.java:
```diff
@@ -1,41 +1,31 @@
-    private static boolean isHoliday(LocalDate date) {
-        return !Holiday.from(date).equals(Holiday.NOT_HOLIDAY);
+    public static boolean isHoliday(LocalDate date) {
+        return Holiday.isHoliday(date.getMonthValue(), date.getDayOfMonth());
```
weekend와 holiday를 나눌 필요가 있을까요?

threepebbles replied:
당장 요구 사항에서는 주말과 공휴일을 묶어서 출석을 하지 않는 날로 사용하고 주말과 공휴일을 구분하고 있지 않아서 주말과 공휴일을 나누지 않아도 요구 사항은 만족할 수 있습니다. 
하지만 주말을 구분하는 기준은 요일이고, 공휴일을 구분하는 기준은 정해진 날짜(월일)이기 때문에 주말과 공휴일을 공통으로 묶어서 관리하기엔 데이터가 일치하는 부분이 없어서 어려운 것 같습니다.
또 추후에 주말 교육 시간이 생겨서 "주말도 출석 기록을 기록한다"로 요구사항이 바뀔 수도 있다는 것을 고려하면 분리해서 관리하는 것이 더 확장성 있겠다고 생각했습니다!

dusdn1702 commented on src/main/java/util/ExceptionHandler.java:
```diff
@@ -2,22 +2,33 @@
-                System.out.println(e.getMessage());
-            }
+        ExecuteResult executeResult;
+        do {
+            executeResult = executeGivenMethod(supplier);
+        } while (!executeResult.isSuccess());
```
```suggestion
        while (!executeResult.isSuccess()) {
            executeResult = executeGivenMethod(supplier);
        } 
```

threepebbles replied:
executeGivenMethod가 실행되기 전까지는 executeResult에 값이 할당되어 있지 않아서 컴파일 에러가 발생할 것 같아요.
line 8에서 excuteResult 선언과 동시에 초기값을 설정해주는 방식으로 해야할까요?

dusdn1702 commented on src/test/java/service/AttendanceServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+import service.dto.AttendanceStatusCount;
+import service.dto.MonthAttendanceStatisticsResponse;
+
+class AttendanceServiceTest {
```
전체적으로 도메인 단위로 TDD 진행해주셨는데 서비스 로직이 핵심일텐데 로직단위 TDD를 진행하지 않으신 이유가 있을까요?
TDD와 테스트 코드의 목적은 무엇일까요?

threepebbles replied:
다시 보다 보니 두 부분에서 서비스 로직에 대한 테스트가 부족해보이네요!

1. 예외에 대한 테스트 부족

예외 상황에 대한 설계가 첫 설계 때 완성된 것이 아니라 구현 중간중간 생각날 때마다 추가되었습니다.

정상적인 입력에 대해 동작하는 기능들에 대해서 먼저 TDD를 하다보니 추후 추가된 예외 상황에 대한 테스트 코드 작성을 깜빡했네요..! 

TDD를 위해서는 `validateCrew()`, `validateOffDay()`, `validateCampusTime()`, `validateSameAttendanceRecordExists()`를 추가할 때마다 각각에 대해 예외를 발생시키는 테스트를 추가해줬어야 했는데, 그러지 못했네요..! 

예외 상황을 추가할 때 TDD를 해야한다는 생각을 잘하지 못한 것 같습니다. 인지하고 연습해보겠습니다.

2. `getRiskCrews()` 테스트 부재

AttendanceService에서 제적 위험 목록 추출 기능(`getRiskCrews()`)에 대한 테스트 코드를 제외하고 작성했는데요. 

`getRiskCrews()` 메서드는 존재하는 크루 목록을 불러와 각 크루에 대해 `getRiskCrew()` 메서드를 통해 제적 위험자 목록을 계산합니다.
`getRiskCrew()`를 구현할 때 월단위 출석 기록 확인 기능(`getMonthAttendanceStatistics()`) 내부에서 사용되는 메서드들과 완전히 동일한 메서드들을 재사용했기 때문에 `getRiskCrews()` 메서드를 작성하는 데는 새로운 도메인 기능의 추가가 하나도 없었습니다. (dto 코드만 추가)

이미 구현된 기존 기능들의 조합으로 새로운 서비스 로직이 순식간에 완성된다는 마음에 테스트를 작성하지 못하고 바로 구현부터 들어갔던 것 같습니다.

내부적으로 동일한 기능을 재사용한다 하더라도 TDD를 해주었어야 하는데 그러지 못했네요..! 의식하면서 하도록 연습해보겠습니다.

dusdn1702 commented on src/test/java/domain/LectureTimeTest.java:
```diff
@@ -1,20 +1,81 @@
+    @DisplayName("해당 날짜가 교육이 있는 날인지 확인한다")
+    void isLectureTimeTest() {
+        // when & then
+        SoftAssertions.assertSoftly(softAssertions -> {
```
SoftAssertions을 사용하신 이유가 무엇인가요~?

threepebbles replied:
SoftAssertions.assertSoftly()를 사용하면 여러 assertThat()에 대한 결과를 한 테스트 메서드에서 확인할 수 있고, 테스트 실패 시 실패한 위치를 알려줍니다. (StackTrace 기능)
비슷한 역할로 Assertions.assertAll()이 있는데 assertAll()은 실패한 테스트에 대해 예외가 발생한 위치를 알려주지 않아서 좀 더 친절한 SoftAssertions을 사용하고 있습니다!

dusdn1702 replied:
오호~ 저도 몰랐네요 감사합니다 👍

dusdn1702 commented:
0 함수를 분리하고 if문을 쓰면 가독성이 좋을 것 같은데요.
가독성이 안좋다고 생각하신 이유가 있을까요?
코스를 진행하면서 요구사항을 충실하게 지켜보는 연습을 하는 것도 좋을 것 같아요.
왜 사용하지 않는게 요구사항이 되었을지, 사용하지 않으면서 구현을 해봤을 때 예상한 것과 일치하는지 등 배울 수 있는 점이 많을 거에요.

1 2 어떻게 TDD를 진행할지에 대해 많은 고민을 하신게 느껴져요.
다만, 코멘트에도 남긴 것처럼 TDD로 얻고 싶은 점이 무엇일지에 대해 생각해보면 좋을 것 같아요.
서비스 로직을 구현하면서 도메인 로직이 변경되지 않을만큼 도메인을 처음부터 잘 설계할 수 있을까요?
서비스 로직 TDD를 진행하면서 필요한 도메인 로직이 생겼을 때 도메인 로직 TDD를 진행해보는 경험도 해보시면 좋을 것 같습니다.

3 이렇게 분리를 해야하는 이유가 무엇일까요? response dto가 모두 service 패키지에 있던데 이렇게 했을 때의 단점은 무엇일지 생각해보면 좋을 것 같아요. 

4 불필요한 값보단 exception이 자연스럽게 느껴졌어요! 👍

5 이상과 초과를 나눠서 표현할 수 있는 방법을 찾아볼까요?

6 리뷰에도 남겼듯 출석 상태라는 도메인을 묶어낼 수 있는 더 좋은 방법이 있을 것 같아요. 이 부분은 한번 고민해봐주시고 다음 피드백 시에도 해결하지 못하셨다면 같이 생각해봐요!

dusdn1702 reviewed (changes requested):
안녕하세요 헤일러~ 2단계 구현 잘 해주셨네요 👏 

남겨주신 질문에 대한 답변과 리뷰 코멘트 남겨두었어요.
궁금한 점 편하게 코멘트 남겨주시면 답변드리도록 할게요!

threepebbles reviewed (commented):
리뷰 감사합니다 샐리!
단일 책임 원칙과 네이밍에 대해 많은 고민이 필요할 것 같네요..!
피드백 해주신 내용 고민해보고 반영해보겠습니다.

threepebbles commented:
dto를 주는 곳과 받는 곳에서 두 곳에서 검증할 일을 한 곳으로 줄인다는 생각으로 dto에서 검증을 진행했는데, 방어적 프로그래밍을 생각하면 dto는 단순히 값만 전달해주고 주는 곳과 받는 곳에서 각각 검증하는 것이 더 신뢰가 높은 코드가 될 것이란 생각이 들었습니다. 검증 로직을 한번 분리해보겠습니다.

threepebbles committed:
refactor(AttendanceController): 초기 화면 기능 실행 로직 switch/case 대신 Map 사용

threepebbles committed:
refactor: 선언한 변수 재사용

threepebbles committed:
refactor(AttendanceStatus): 상수 값 대신 람다 함수 사용

threepebbles committed:
refactor(AttendanceRecord): 책임에 맞는 위치로 검증 메서드 이동

threepebbles committed:
refactor(AttendanceRecord): 생성자 추가 및 적용

threepebbles committed:
feat: repository 대신 일급 컬렉션 추가(Crews, AttendanceRecords)

* CrewRepository -> Crews
* AttendanceRecordRepository -> AttendanceRecords

threepebbles committed:
refactor: 변수 선언 한 줄로 통합

threepebbles committed:
refactor: 모든 공휴일 검증은 Holiday 클래스에 위임하도록 수정

threepebbles committed:
fix(AttendanceRecords): 생성자에서 출석 기록을 추가하도록 수정

threepebbles committed:
feat(Validator): null 검증 기능 추가

threepebbles committed:
feat(AbstractAttendanceRecord): 출석 기록 객체 추상화 및 기존 기능 제거

- 출석 기록 저장 기능 리팩터링
- 출석 기록 수정 기능 제거
- 크루별 월단위 출석 기록 확인 기능 제거
- 제적 위험자 목록 조회 기능 제거

threepebbles committed:
refactor(AttendanceRecords): 출석 기록 저장소 자료구조 변경(List -> Set)

threepebbles committed:
fix(AttendanceRecords): find 파라미터 수정

threepebbles committed:
feat: 출석 기록 수정 기능 추가

threepebbles committed:
refactor(AttendanceStatus): 예외 메시지 수정

threepebbles committed:
refactor(RiskRank): 상수 값 대신 람다 메서드 사용

threepebbles committed:
refactor(AttendanceRecords): 출석 기록 저장소 자료구조 수정(Set -> List)

threepebbles committed:
feat(EmptyAttendanceRecord): 생성자 수정, 정적 팩터리 메서드 추가

threepebbles committed:
feat: 월단위 출석 기록 확인 기능 추가

threepebbles committed:
test(DateTimeUtil): 일관된 프로그램 실행 결과를 위해 nowDate()을 25년 2월 28일로 고정

threepebbles committed:
feat: 제적 위험자 목록 조회 기능 추가

threepebbles committed:
refactor: 변수 추출

threepebbles committed:
refactor: 줄간격 정리

threepebbles committed:
remove: 안쓰는 메서드 제거

threepebbles committed:
refactor: 불변 필드에 final 키워드 추가

threepebbles committed:
refactor: fixture 패키지 분리

threepebbles committed:
refactor: 테스트 메서드 네이밍 일관되게 수정

threepebbles committed:
fix: 수정 전 출석 기록이 예외 발생하지 않도록 수정

threepebbles committed:
refactor: DateTimeFomatter 위치 이동

threepebbles committed:
refactor(Command): Map 대신 Consumer 사용

threepebbles committed:
fix: 제적 위험 등급 출력 양식에 맞게 출력

threepebbles committed:
refactor: 메서드 재사용하도록 수정

threepebbles committed:
refactor: DateUtil에 정의된 DateTimeFormatter 이용하도록 수정

threepebbles committed:
refactor: 의미가 드러나도록 메서드명 수정

threepebbles committed:
feat(AbstractAttendanceRecord): equals 재정의 제거

threepebbles committed:
refactor: 줄바꿈 수정

threepebbles committed:
refactor: 파일 입출력 방식 변경

threepebbles committed:
refactor: 줄바꿈 수정, 변수 네이밍 변경

threepebbles committed:
refactor: 매직넘버 상수화

threepebbles committed:
refactor: 날짜 <-> 문자열 변환 책임 DateTimeUtil로 위임

threepebbles committed:
refactor: 줄바꿈 수정

threepebbles committed:
refactor: 관련있는 dto끼리 내부 클래스로 묶어서 관리

threepebbles committed:
test: DateUtil 테스트 추가

threepebbles committed:
refactor(AttendanceService): 메서드들을 책임에 맞는 클래스로 분리

threepebbles committed:
refactor: riskCount 검증 위치 수정

threepebbles committed:
remove: 사용하지 않는 메서드 제거

threepebbles committed:
refactor: 메서드를 책임에 맞는 클래스로 분리

threepebbles committed:
refactor(RiskRank): 위험 등급 가중치 계산 기능 캡슐화

threepebbles committed:
refactor: 테스트 예제(given) 리팩터링

threepebbles committed:
test: 공휴일 검증 테스트 추가

threepebbles committed:
refactor: 테스트 메서드 분리

threepebbles committed:
refactor: 테스트명 명확하게 수정

threepebbles committed:
refactor: 테스트 픽스처 사용하도록 수정

threepebbles commented:
안녕하세요 샐리, 저번에 주신 피드백 반영하고 리팩터링 해봤어요!

피드백 반영하면서 기존 설계가 문제가 많은 것 같아서 대부분의 코드를 주석 처리하고 처음부터 TDD를 다시 해봤어요.
기능 구현을 끝낸 후 프로덕션 코드를 리팩터링하면서 메서드가 분리되거나, 합쳐지거나, 클래스에 대한 책임이 바뀌어 다른 클래스로 이전될 때마다 테스트 코드를 동시에 추가하고 수정해주어야 하는 점이 조금 헷갈리고 어려웠던 것 같아요.
나름 집중해서 TDD를 한다고 하긴 했지만 검토하면서 보니까 그래도 빠뜨린 테스트들이 몇몇 있더군요 😢 
이번에도 리뷰 잘 부탁드립니다 🙂 

## 주요 변경 사항
### 1️⃣ 크루 목록, 출석 기록 관리 일급 컬렉션 사용
일급 컬렉션에게 목록 조회, 삭제에 대한 책임을 부여하도록 수정했습니다.

### 2️⃣ 기록이 없는 결석에 대한 처리
기록이 있는 결석과 기록이 없는 결석을 어떻게 처리해야 할 지 고민이 많았던 것 같아요.
기록이 없는 결석의 LocalTime에 의미 없는 값("00:00", "23:59")을 넣어놓거나 null 혹은 Optional을 넣어놓는 것은 차후 확장과 변경에 대해 유연하지 못한 설계라는 생각이 들어서 추상화 클래스와 다운 캐스팅을 시도해봤어요.

### 3️⃣ 제적 위험자 목록 정렬 위치 변경
제적 위험자 목록이 제대로 정렬됐는지 테스트해보고 싶어서 dto에 담으면서 같이 정렬했어요.

---

## ❓ 질문
### 1️⃣ 테스트 메서드명 컨벤션
이번 미션에서는 테스트 메서드를 아래와 같이 프로덕션 코드의 실제 메서드명과 동일하게 하다보니, findRiskCrews()가 아니라 getRiskCrews()로 네이밍이 바뀌면 테스트 코드도 따라가서 바꿔줘야 하는 번거러움이 있었습니다.
그래서 다음부터는 DisplayName에 언더바만 붙힌 한글 테스트명을 사용할까 고민중입니다. 
샐리는 테스트 메서드명을 어떻게 짓는 걸 선호하시나요?
```java
// 프로덕션 코드.java
public List<RiskCrew> findRiskCrews() {
    ...
}

// 테스트 코드.java
@Test
@DisplayName("정렬된 제적 위험자 목록을 조회한다.")
void findRiskCrews_test() {
    ...
}
```

threepebbles commented:
안녕하세요 샐리, 저번에 주신 피드백 반영하고 리팩터링 해봤어요!

피드백 반영하면서 기존 설계가 문제가 많은 것 같아서 대부분의 코드를 주석 처리하고 처음부터 TDD를 다시 해봤어요.
기능 구현을 끝낸 후 프로덕션 코드를 리팩터링하면서 메서드가 분리되거나, 합쳐지거나, 클래스에 대한 책임이 바뀌어 다른 클래스로 이전될 때마다 테스트 코드를 동시에 추가하고 수정해주어야 하는 점이 조금 헷갈리고 어려웠던 것 같아요.
나름 집중해서 TDD를 한다고 하긴 했지만 검토하면서 보니까 그래도 빠뜨린 테스트들이 몇몇 있더군요 😢 
이번에도 리뷰 잘 부탁드립니다 🙂 

## 주요 변경 사항
### 1️⃣ 크루 목록, 출석 기록 관리 일급 컬렉션 사용
일급 컬렉션에게 목록 조회, 삭제에 대한 책임을 부여하도록 수정했습니다.

### 2️⃣ 기록이 없는 결석에 대한 처리
기록이 있는 결석과 기록이 없는 결석을 어떻게 처리해야 할 지 고민이 많았던 것 같아요.
기록이 없는 결석의 LocalTime에 의미 없는 값("00:00", "23:59")을 넣어놓거나 null 혹은 Optional을 넣어놓는 것은 차후 확장과 변경에 대해 유연하지 못한 설계라는 생각이 들어서 추상화 클래스와 다운 캐스팅을 시도해봤어요.

### 3️⃣ 제적 위험자 목록 정렬 위치 변경
제적 위험자 목록이 제대로 정렬됐는지 테스트해보고 싶어서 dto에 담으면서 같이 정렬했어요.

---

## ❓ 질문
### 1️⃣ 테스트 메서드명 컨벤션
이번 미션에서는 테스트 메서드를 아래와 같이 프로덕션 코드의 실제 메서드명과 동일하게 하다보니, findRiskCrews()가 아니라 getRiskCrews()로 네이밍이 바뀌면 테스트 코드도 따라가서 바꿔줘야 하는 번거로움이 있었습니다.
그래서 다음부터는 DisplayName에 언더바만 붙힌 한글 테스트명을 사용할까 고민중입니다. 
샐리는 테스트 메서드명을 어떻게 짓는 걸 선호하시나요?
```java
// 프로덕션 코드.java
public List<RiskCrew> findRiskCrews() {
    ...
}

// 테스트 코드.java
@Test
@DisplayName("정렬된 제적 위험자 목록을 조회한다.")
void findRiskCrews_test() {
    ...
}
```

threepebbles committed:
refactor(AttendanceRecordsGenerator): 출석 기록 생성기 리팩터링

threepebbles committed:
refactor: 메서드명 수정

threepebbles committed:
refactor(InputView): DateTimeUtil 상수 사용하도록 수정

threepebbles committed:
refactor: 메서드명 변경

threepebbles committed:
refactor: 메서드명 변경

threepebbles committed:
refactor(CampusTimePolicy): 캠퍼스 시간 정책 분리

threepebbles committed:
refactor: actual, expected 위치 변경

dusdn1702 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -1,66 +1,86 @@
-        do {
-            isRunning = executeMainMenu();
-        } while (isRunning);
+        while (isRunning) {
+            Command command = Command.from(InputView.scanMenuCommand());
+            command.run(this);
```
👍

dusdn1702 commented on src/main/java/controller/Command.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.function.Consumer;
+import util.ExceptionHandler;
+
+public enum Command {
```
switch문에서 이렇게 enum으로 변경한 이유가 있나요?
이렇게 구현했을 때 어떤 점이 좋았고, 안좋았는지 궁금하네요 😄

threepebbles replied:
enum에서 커맨드와 관련된 정보만 포함하고 있고, 각 입력(`String`)에 대해에 대해 어떤 행위(`Function`)를 할 것인지 묶어서 볼 수 있어서 응집도가 높아졌다고 생각해요!

안 좋은 점은 command와 controller가 서로 양방향으로 의존성이 생겨서 강결합이 되어 있다는 것일까요? 큰 단점인지는 잘 모르겠어요!

dusdn1702 replied:
저도 좋은 방향이라고 생각해요~ 
다만 말씀 주신 것처럼 Command 내부에서 Controller 메서드를 가져다 쓰는 부분이 의존성이 커지긴 하지만 지금 구조에서는 최선일 것 같네요.
이전에 switch 문이 더 깔끔하다고 느꼈다고 말씀 주셔서 어떻게 느끼셨나 궁금했어요! 👍

dusdn1702 commented on src/main/java/domain/AbstractAttendanceRecord.java:
```diff
@@ -0,0 +1,54 @@
+import util.DateTimeUtil;
+import util.Validator;
+
+public abstract class AbstractAttendanceRecord {
```
EmptyAttendanceRecord 를 위해 추상화를 해주신 것 같은데 어땠나요~? 오히려 더 복잡한 구조가 되진 않았나요? LocalTime을 객체화해서 사용할 수도 있었을 것 같아요.

threepebbles replied:
"기록이 없는 출석 기록(현재 EmptyAttendanceRecord)"의 LocalTime 필드에 null을 사용하면 NullPointException이 걱정되고, Optional을 사용하면 NoSuchElementException이 걱정되고, 의미 없는 값("00:00", "23:59")을 사용하는 것은 오해의 여지를 남길  수 있는 코드라고 생각했어요.
추상화 하게 된 이유는 LocalTime의 상태를 null, Optional, 의미 없는 값("00:00", "23:59")이 아닌 의미 있는 값을 갖게 묶을 다른 방법이 도저히 생각나지 않아서에요. (현재 아는 지식 내에서는 이게 최선이라 생각했습니다.) 
LocalTime을 객체화해도 같은 문제가 발생하지 않나요?

이런 경우에 그냥 null을 사용하고 null을 꼼꼼하게 체크해서 NPE의 가능성을 줄이는 방향으로 설계하는 게 더 나은 방법일까요?

dusdn1702 replied:
추상화는 양날의 검이라고 생각해요.
이 코드를 보고 EmptyAttendanceRecord가 기록이 없는 출석이라고 생각하기는 쉽지 않아보이거든요. 추후에 다른 AbstractAttendanceRecord의 구현체가 나올 가능성도 없다고 볼 수 있죠.
LocalTime을 객체화하고 NullPointException나 NoSuchElementException을 객체 내부에서 대응할 수 있도록 구현할 수도 있겠죠?
하지만 이렇게 다양한 시도를 해보고 미션을 진행하시면서 헤일러의 기준을 찾아나가면 좋을 것 같아요!

dusdn1702 commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -1,10 +1,47 @@
+        return time;
+    }
+
+    public String getNickname() {
```
사용처가 없는 것 같아요!

dusdn1702 commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,82 @@
+        attendanceRecords.add(attendanceRecord);
+    }
+
+    public void overwriteAttendanceRecord(AttendanceRecord after) {
```
일관된 네이밍을 하면 어떨까요? add는 추가인데 이 함수는 수정을 하면서 overwriteAttendanceRecord까지 붙네요.

dusdn1702 commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,82 @@
+                        && date.equals(record.getDate()));
+    }
+
+    public AttendanceRecord getOneByCrewAndDate(Crew crew, LocalDate date) {
```
여러개 중 하나의 값을 가져오는 건 아니지 않나요? 크루와 날짜로 조회하면 하나의 값인데 One이 붙는 이유가 궁금해요.

threepebbles replied:
네이밍 규칙을 헷갈린 것 같습니다 😅 수정했습니다

dusdn1702 commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public List<AbstractAttendanceRecord> getByCrewFromTo(Crew crew, LocalDate from, LocalDate to) {
+        return Stream.iterate(from, date -> date.plusDays(1))
```
iterate을 사용한 이유가 있을까요? for문과 비교했을 때 헤일러는 어떤게 가독성이 좋다고 생각하나요?

threepebbles replied:
호출되는 메서드명(limit, filter, map)을 보고 기능을 빠르게 짐작할 수 있기 때문에 iterate을 사용하는 것이 가독성이 더 좋다고 생각했습니다.
limit에서 리스트의 크기를 짐작할 수 있고, filter에서 어떤 조건으로 걸러졌는지 알 수 있고, map에서 어떤 데이터 타입으로 변환됐는지 알 수 있습니다.

for문을 사용하면 아래와 같은 코드가 됩니다.
```java
List<AbstractAttendanceRecord> abstractAttendanceRecords = new ArrayList<>();
for(LocalDate date = from; date.isBefore(to.plusDays(1));date = date.plusDays(1)) {
    if(!LectureTime.isLectureDate(date)) {
        continue;
    }
    abstractAttendanceRecords.add(findByCrewAndDate(crew, date));
}
return abstractAttendanceRecords;
```
우선 indent 깊이 1 제약사항을 지킬 수 없고, abstractAttendanceRecords 변수를 선언 해주어야 해서 코드의 길이가 전체적으로 늘어났습니다.

dusdn1702 commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,82 @@
+
+    public Map<AttendanceStatus, Integer> getAttendanceStatistics(Crew crew, LocalDate from, LocalDate to) {
+        List<AbstractAttendanceRecord> records = getByCrewFromTo(crew, from, to);
+        Map<AttendanceStatus, Integer> statusCount = new HashMap<>();
```
Map<AttendanceStatus, Integer> statusCount 도 일급컬렉션 분리해보면 어떨까요?

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,37 @@
-    ABSENT_LATE("결석", "출석 시간 30분 초과해서 결석", 30),
-    ABSENT("결석", "출석 기록을 하지 않아서 결석", Integer.MAX_VALUE),
-    OFF_DAY("쉬는 날", "출석 상태를 정의할 수 없음", Integer.MAX_VALUE),
+    ATTENDANCE("출석", (elapsedMinutes) -> elapsedMinutes <= 5),
+    LATE("지각", (elapsedMinutes) -> 5 < elapsedMinutes && elapsedMinutes <= 30),
```
이 사이에 10분 초과 시 지각으로 변경되었는데 하나의 값만 수정하는 실수가 나오기 쉽겠네요.

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -1,44 +1,37 @@
-    private final String title;
     private final String description;
-    private final int elapsedMinutesLimit;
+    private final Function<Integer, Boolean> condition;
```
이 condition 값은 이 AttendanceStatus 에서만 사용되는데요, enum에서 멤버변수로 가질 필요가 있을까요? 함수로 가져가도 되지 않을까요?

dusdn1702 commented on src/main/java/domain/RiskRank.java:
```diff
@@ -0,0 +1,62 @@
+                .orElseThrow(() -> new IllegalStateException("논리적으로 도달할 수 없는 예외입니다."));
+    }
+
+    public static RiskRank of(int lateCount, int absentCount) {
```
테스트를 위한 프로덕션 코드를 생성하는 것에 대해서 어떻게 생각하시나요?

threepebbles replied:
초기에는 of를 프로덕션 코드에서 사용하게 해서 테스트를 작성했다가, from으로 대체하게 되면서 지우지 않았네요..! 테스트도 from 테스트로 수정했습니다.

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,81 @@
-        return new ModifiedResult(request.date(), before, after);
+    public SaveAttendanceRecordResponse saveAttendanceRecord(SaveAttendanceRequest request) {
+        Crew crew = crews.findByNickname(request.nickname());
+        AttendanceRecord saved = AttendanceRecord.of(crew, request.date(), request.time());
```
save 된 적이 있나요?? 왜 saved일까요?

threepebbles replied:
네이밍을 잘못했네요!

혹시 이번 피드백과 별개로 질문드리고 싶은것이 있는데요, 데이터를 저장하고 저장된 데이터를 반환해주어야 할 때, 아래의 두 가지를 방법을 생각해봤어요.
1️⃣ 저장할 때 사용한 객체(willBeSaved)를 그대로 반환
```java
public SaveAttendanceRecordResponse saveAttendanceRecord(SaveAttendanceRequest request) {
    Crew crew = crews.findByNickname(request.nickname());
    AttendanceRecord willBeSaved = AttendanceRecord.of(crew, request.date(), request.time());
    attendanceRecords.addIfAbsent(willBeSaved);
    return SaveAttendanceRecordResponse.of(willBeSaved);
}
```
2️⃣ 저장할 때 사용한 객체가 아닌 조회를 통해 찾아진 값(`found`)를 넘겨주는 방식
```java
public SaveAttendanceRecordResponse saveAttendanceRecord(SaveAttendanceRequest request) {
    Crew crew = crews.findByNickname(request.nickname());
    AttendanceRecord willBeSaved = AttendanceRecord.of(crew, request.date(), request.time());
    attendanceRecords.addIfAbsent(willBeSaved);
    AttendanceRecord found = attendanceRecords.getByCrewAndDate(crew, request.date());
    return SaveAttendanceRecordResponse.of(found);
}
```
2번 방식은 조회라는 연산을 한번 더 수행하는 대신 조회를 통해 확실하게 저장되었음을 알 수 있는 장점이 있는 것 같아요.
1번 방식은 연산의 낭비는 없지만 `addIfAbsent()` 메서드에 대해 온전히 신뢰하는 코드에요.
신뢰와 비용을 고려했을 때 어느 방식을 사용하는 게 더 좋은 선택일까요?

dusdn1702 replied:
1의 경우는 저장된 객체가 willBeSaved와 동일하다고 보장할 수 있나요? 실제 저장하면서 id가 추가되는 방식이라면 어떻게 될까요?
음 추후에 학습하겠지만 JpaRepository 의 save 메서드를 한번 미리 읽어보면 고민이 조금은 해결 될 것 같아요 👍

dusdn1702 commented on src/main/java/service/AttendanceService.java:
```diff
@@ -1,111 +1,81 @@
+    public MonthAttendanceStatisticsResponse getMonthAttendanceStatistics(MonthAttendanceStatisticsRequest request) {
+        Crew crew = crews.findByNickname(request.nickname());
+        LocalDate from = request.today().withDayOfMonth(1);
+        LocalDate to = request.today().minusDays(1);
```
AttendanceRecords#getByCrewFromTo와 여기서 1을 더하고 빼는 작업이 많아서 실수하기가 쉬워보여요.

dusdn1702 commented on src/test/java/domain/AttendanceRecordTest.java:
```diff
@@ -0,0 +1,66 @@
+
+public class AttendanceRecordTest {
+
+    @Nested
```
Nested 를 사용한 이유가 무엇인가요?
예외가 발생하지 않는 테스트가 사실은 핵심적은 비즈니스 로직 테스트 아닌가요? 이렇게 나눌때의 장점이 있을까요?

threepebbles replied:
정상적인 입력을 기준으로 TDD를 먼저 진행하고, 예외 입력에 대한 검증은 중간에 계속 추가되었던 것 같습니다.
그래서 예외 상황이 추가될 때마다 예외 상황만 구분해서 테스트 성공 여부를 확인하고 싶었습니다. 
정상적인 입력에 대해 잘 동작한다는 것과, 예상된 예외 상황을 잘 잡아내는가를 분리해서 보는 것이 가독성도 더 높아질 것이라 생각했습니다.

dusdn1702 commented on src/test/java/domain/AttendanceRecordTest.java:
```diff
@@ -0,0 +1,66 @@
+            // given
+            String nickname = "우가";
+            Crew crew = new Crew(nickname);
+            LocalDate monday = LocalDate.of(2025, 2, 3);
```
월요일이 출석이 아닌 날로 변경되면 이렇게 되어 있는 값들을 모두 찾아 바꿔주어야겠네요. 원하는 테스트는 정상일에 대한 검증이므로 확실하게 그런 값을 사용하는건 어떨까요?

dusdn1702 commented on src/test/java/domain/RiskRankTest.java:
```diff
@@ -0,0 +1,47 @@
+class RiskRankTest {
+    private static Stream<Arguments> ofCases() {
+        return Stream.of(
+                Arguments.of(0, 0, RiskRank.NOT_MANAGED),
+                Arguments.of(2, 1, RiskRank.NOT_MANAGED),
+                Arguments.of(0, 2, RiskRank.WARNING),
+                Arguments.of(2, 2, RiskRank.WARNING),
+                Arguments.of(0, 3, RiskRank.INTERVIEW),
+                Arguments.of(0, 4, RiskRank.INTERVIEW),
+                Arguments.of(0, 5, RiskRank.INTERVIEW),
+                Arguments.of(0, 6, RiskRank.EXPELLED)
```
👍 👍

dusdn1702 commented on src/test/java/service/AttendanceServiceTest.java:
```diff
@@ -0,0 +1,173 @@
+            AttendanceService attendanceService = new AttendanceService(crews, new AttendanceRecords(allRecords));
+            RiskCrewsRequest request = new RiskCrewsRequest(to.plusDays(1));
+
+            // 제적 위험자 순서: 부기, 미소, 우가, 포스티
+            // when & then
+            RiskCrewsResponse response = attendanceService.findRiskCrews(request);
```
이 순서는 어떻게 보장이 되나요?
리팩터링하며 순서가 바뀌면 다 실패하는 테스트네요.
그럼 프로덕션 코드는 바뀐 것 없이 테스트를 고쳐야하는데 우리가 원하는 테스트 코드의 이상향이 맞을까요?

threepebbles replied:
순서가 바뀐다고 함은 정렬 기준에 대한 요구사항이 바뀌는 경우인가요?

threepebbles replied:
순서가 바뀐다고 함은 정렬 기준에 대한 요구사항이 바뀌는 경우인가요?

threepebbles replied:
리팩터링하며 순서가 바뀌는 경우가 어떤 경우가 있을까요?

dusdn1702 replied:
만약 이 로직이 확장이 필요해서 10명의 크루를 관리해야한다면 10명의 위험 순서를 다 관리하고 알아볼 수 있을까요?
미소의 결석 수를 하나만 실수로 늘려도 테스트는 실패할텐데요. 테스트의 유지보수가 프로덕션 로직보다 어려울 것 같아요.

dusdn1702 commented on src/test/java/service/AttendanceServiceTest.java:
```diff
@@ -0,0 +1,173 @@
+
+        @Test
+        @DisplayName("정렬된 제적 위험자 목록을 조회한다.")
+        void findRiskCrews_test() {
```
이것보단 DisplayName에 언더바만 붙힌 한글 테스트명을 사용하는 것이 좋을 것 같아요 😄 
여기서 예외 테스트까지 추가한다면 어떻게 되나요? findRiskCrews_exception_test ?? ㅎㅎ
저는 보통 테스트의 목적을 가진 메서드명을 붙이는 것 같아요~

dusdn1702 reviewed (changes requested):
안녕하세요 헤일러~
짧은 시간이었을텐데도 많은 고민을 통해 개선과 그 안에서도 TDD를 수행해주신 점이 인상 깊었어요.
고생 많으셨습니다. 
간단한 피드백 몇가지 남겨두었어요. 확인 부탁드릴게요!

threepebbles commented:
outdated 되어서 외부 커멘트로 답변하겠습니다!
https://github.com/woowacourse/java-attendance/pull/84#discussion_r1976562994
> static final 키워드를 붙인 경우 상수라고 취급할 수 없을까요? 요걸 잘 이해하지 못했어요! 어떤 의미인가요?

static final 키워드가 붙으면 전부 상수 취급해야 한다고 생각했었는데 그게 아니었네요..! 변경 가능한(mutable) 값이 있으면 상수가 아니었군요.
[reference](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)
```java
// constant
static final SomeMutableType[] EMPTY_ARRAY = {};
static final Map<String, Integer> AGES = ImmutableMap.of("Ed", 35, "Ann", 32);

// not constant
static final Set<tring> mutableCollection = new HashSet<String>();
static final ImmutableSet<SomeMutableType> mutableElements = ImmutableSet.of(mutable);
```

threepebbles committed:
fix: 공휴일 테스트 수정

threepebbles committed:
refactor: 출석 기록 존재 여부를 확인하는 추상 메서드 추가

threepebbles committed:
refactor: 메서드명 수정

threepebbles committed:
refactor(AttendanceStatusCount): 클래스 분리

threepebbles committed:
feat: 교육이 있는 날인지 검증 추가

threepebbles committed:
remove: 사용하지 않는 메서드 제거

threepebbles committed:
refactor: 달의 첫 날과 전날을 계산하는 로직 캡슐화

threepebbles committed:
refactor: LocalDate과 LocalTime은 테스트 픽스처를 사용해서 테스트 하도록 수정

threepebbles committed:
test(findRiskCrews): 정렬 순서가 아닌 제적 위험 등급을 검증하도록 수정

threepebbles committed:
test: 제적 위험자 정렬 순서 검증 추가

threepebbles commented:
안녕하세요 샐리! 주신 피드백 다시 반영해보았습니다.
https://github.com/woowacourse/java-attendance/pull/84#discussion_r1976570496 
해당 커멘트에서 리팩터링이 어떻게 일어나면 테스트 결과가 바뀔 수 있을지, 
제가 코드에서 놓친 부분이 있을지 잘 모르겠습니다..!

dusdn1702 reviewed (approved):
안녕하세요 헤일러~
리뷰 반영해주신 부분들 확인했습니다! 빠르게 잘 수정해주셨네요 👏 
이번 미션은 여기서 마무리 해도 될 것 같아서 머지하도록 할게요 
고생 많으셨습니다 👍
